File: MaxCounters/Python/MaxCounters.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(N, A):
    res = [0] * N
    maxim = 0
    increase = 0
    M = len(A)
    # for every element in array size
    for x in range(M):
        # if its single incrementation
        if A[x] > 0 and A[x] <= N:
            # if there needs has been max command before not applied
            if res[A[x] - 1] < increase:
                res[A[x] - 1] = increase
            # increment
            res[A[x] - 1] += 1
            # if its biggest one on array
            if res[A[x] - 1] > maxim:
                maxim = res[A[x] - 1]
        # max command, save new value to increase
        if A[x] == N + 1:
            increase = maxim
    # loop to apply every non applied max increases
    for i in range(0, N):
        if res[i] < increase:
            res[i] = increase
    return res

File: MaxCounters/C/MaxCounters.c

// two possible operations:
// either count up a[X]++;
// or set all from a[0] to a[N] = max counter;
// i want to do it simple way by making 2 different functions, one for each

// order of operation
// 1. allocate for result structure, size of times N
// 2. Check first digit and go outside to right function
// 3. iterate the whole array and repeat second step.

// N = Number of arrays needed, A = Array, M = counter number

//EDIT 1: THIS DOESNT WORK IF MAX IS 0 BEFORE 1st count
// EDIT 2: IM dumb, max should not be pumped up if there is nothing that pumps it up in the array

//result is  88%, there is still some kind of timeout with large

// now 100%, modified helper function to while loop and only use it once after primary loop.


struct Results {
  int * C;
  int L; // Length of the array
};

// you can write to stdout for debugging purposes, e.g.
// printf("this is a debug message\n");


struct Results solution(int N, int A[], int M) {
    // Implement your solution here
    struct Results result;
    int *res;
    int x = 0;
    int max = 0;
    int increase = 0;

    res = (int *)calloc(N + 1, sizeof(int));

    while (x < M + 1)
    {
//        printf("Current A[%d]: %d \n", x, A[x]);
        if (A[x] > 0 && A[x] < N + 1)
        {
            if (res[A[x] - 1] < increase)
                res[A[x] - 1] = increase;
            res[A[x] - 1]++;
            if (res[A[x] - 1] > max)
                max++;
        }
        if (A[x] == N + 1)
        {
            increase = max;
        }
        x++;
    }
    x = 0;
    while (x < N)
    {
        if (res[x] < increase)
            res[x] = increase;
//        printf("RES[%d]: %d \n", x, res[x]);
        x++;
    }
    res[x] = '\0';
    result.C = res;
    result.L = N;
    return result;
}